File: solutions/java/poker/1/src/main/java/Poker.java
import java.util.*;
import java.util.stream.*;
import java.util.function.*;

class Poker {
    private final List<String> bestHands;
    Poker(List<String> hands) {
        List<Hand> evaluated = hands.stream().map(Hand::new).toList();
        Hand bestHand = Collections.max(evaluated);
        bestHands = evaluated.stream().filter(h -> h.compareTo(bestHand) == 0).map(Hand::toString).toList();
    }

    List<String> getBestHands() {
        return bestHands;
    }

}

class Hand implements Comparable<Hand> {
    final String handStr;
    final List<Card> cards;

    public Hand(String handStr) {
        this.handStr = handStr;
        cards = Arrays.stream(handStr.split(" "))
            .map(s -> {
                int splitPoint = s.length() - 1;
                String rankStr = s.substring(0, splitPoint);
                int rank = switch (rankStr) {
                    case "J" -> 11;
                    case "Q" -> 12;
                    case "K" -> 13;
                    case "A" -> 14;
                    default -> Integer.parseInt(rankStr);
                };
                return new Card(rank, s.charAt(splitPoint));
            }).toList();
    }

    @Override
    public int compareTo(Hand otherHand) {
        int thisRanking = Evaluator.getRanking(this.cards);
        int otherRanking = Evaluator.getRanking(otherHand.cards);
        if (thisRanking != otherRanking) return Integer.compare(otherRanking, thisRanking);
        
        List<Integer> thisKickerList = this.getKickerList();
        List<Integer> otherKickerList = otherHand.getKickerList();
        for (int i = 0; i < thisKickerList.size(); i++) {
            int thisRank = thisKickerList.get(i);
            int otherRank = otherKickerList.get(i);
            if (thisRank != otherRank) return Integer.compare(thisRank, otherRank);
        }
        return 0;
    }
    
    private List<Integer> getKickerList() {
        List<Card> sorted = new ArrayList<>(this.cards);
        Collections.sort(sorted);
        boolean isAValue1 = Evaluator.isConsecutive(sorted) && sorted.getFirst().rank == 2;
        
        Map<Integer, Integer> frequency = Evaluator.getRankFrequency(sorted);
        return frequency.entrySet()
            .stream()
            .map(e -> isAValue1 && e.getKey() == 14 ? Map.entry(1, e.getValue()) : e)
            .sorted(Comparator.comparing(e -> ((Map.Entry<Integer, Integer>) e).getValue())
                .thenComparing(e -> ((Map.Entry<Integer, Integer>) e).getKey())
                .reversed())
            .map(Map.Entry::getKey)
            .distinct()
            .toList();
    }

    @Override
    public String toString() {
        return handStr;
    }
}

class Card implements Comparable<Card> {
    final int rank;
    final char suit;
    public Card(int rank, char suit) {
        this.rank = rank;
        this.suit = suit;
    }

    @Override
    public int compareTo(Card otherCard) {
        return Integer.compare(this.rank, otherCard.rank);
    }
}

class Evaluator {
    private static final List<Predicate<List<Card>>> RANKINGS = List.of(
        cards -> cards.get(0).rank == 10 && isSameSuit(cards) && isConsecutive(cards),
        cards -> isSameSuit(cards) && isConsecutive(cards),
        cards -> isSameRank(cards, 4),
        cards -> isSameRank(cards, 3, 2),
        cards -> isSameSuit(cards),
        cards -> isConsecutive(cards),
        cards -> isSameRank(cards, 3),
        cards -> isSameRank(cards, 2, 2),
        cards -> isSameRank(cards, 2)
    );

    private static boolean isSameSuit(List<Card> cards) {
        return getSuits(cards).distinct().count() == 1;
    }

    private static boolean isSameRank(List<Card> cards, int... counts) {
        IntStream ranks = getRanks(cards);
        if (counts.length == 0) return ranks.distinct().count() == 1;
        Map<Integer, Integer> frequency = getRankFrequency(cards);
        for (int count : counts) {
            Integer matchingRank = frequency.entrySet()
                .stream()
                .filter(e -> e.getValue() >= count)
                .reduce((e1, e2) -> e1.getValue() > e2.getValue() ? e2 : e1)
                .orElse(Map.entry(0, 0))
                .getKey();
            if (frequency.remove(matchingRank) == null) return false;
        }
        return true;
    }

    public static boolean isConsecutive(List<Card> cards) {
        int[] ranks = getRanks(cards).distinct().toArray();
        int length = ranks.length;
        if (length != cards.size()) return false;
        if (ranks[0] == 2 && ranks[length - 2] == 5 && ranks[length - 1] == 14) return true;
        return ranks[length - 1] - ranks[0] == length - 1;
    }

    private static IntStream getRanks(List<Card> cards) {
        return cards.stream().mapToInt(card -> card.rank);
    }
    
    private static IntStream getSuits(List<Card> cards) {
        return cards.stream().mapToInt(card -> card.suit);
    }
    
    public static Map<Integer, Integer> getRankFrequency(List<Card> cards) {
        return getRanks(cards).boxed().collect(Collectors.toMap(k -> k, v -> 1, Integer::sum));
    }
    
    public static int getRanking(List<Card> cards) {
        List<Card> sorted = new ArrayList<>(cards);
        Collections.sort(sorted);
        for (int i = 0; i < RANKINGS.size(); i++) {
            if (RANKINGS.get(i).test(sorted)) return i + 1;
        }
        return 10 + 14 - sorted.getLast().rank;
    }
}

File: solutions/python/poker/3/poker.py
from collections import Counter


LETTER_RANKS = {"J": 11, "Q": 12, "K": 13, "A": 14}

RANKINGS = (
    lambda ranks, suits: consecutive(ranks) and same(suits),
    lambda ranks, suits: same(ranks, 1, 4),
    lambda ranks, suits: same(ranks, 2, 3),
    lambda ranks, suits: same(suits),
    lambda ranks, suits: consecutive(ranks),
    lambda ranks, suits: same(ranks, 1, 1, 3),
    lambda ranks, suits: same(ranks, 1, 2, 2),
    lambda ranks, suits: same(ranks, 1, 1, 1, 2)
)

def same(objects, *counts: int) -> bool:
    if not counts: counts = (len(objects),)
    return list(counts) == sorted(Counter(objects).values())

def consecutive(ranks: list[int]) -> bool:
    length = len(ranks)
    if length != len(set(ranks)): return False
    return ranks[-1] - ranks[0] == length - 1

    
class Hand:

    def __init__(self, hand_str: str):
        self.hand_str = hand_str
        splitted = hand_str.split(" ")
        ranks, suits = [s[:-1] for s in splitted], [s[-1] for s in splitted]
        self.ranks = sorted(LETTER_RANKS[s] if s in LETTER_RANKS else int(s) for s in ranks)

        if self.ranks[-1] == 14:
            a_is_1 = [1] + self.ranks[:-1]
            if consecutive(a_is_1):
                self.ranks = a_is_1
        
        self.suits = list(suits)

    @property
    def ranking(self) -> int:
        for i, test in enumerate(RANKINGS):
            if test(self.ranks, self.suits): return i
        return len(RANKINGS)

    @property
    def kicker(self) -> list[int]:
        return [t[0] for t in Counter(reversed(self.ranks)).most_common()]

    def __lt__(self, other) -> bool:
        return self.compare_to(other) < 0

    def compare_to(self, other) -> int:
        if self.ranking != other.ranking:
            return other.ranking - self.ranking
        return (self.kicker > other.kicker) - (self.kicker < other.kicker)


def best_hands(hands: list[str]) -> list[str]:
    processed = [Hand(s) for s in hands]
    best_hand = max(processed)
    return [h.hand_str for h in processed if best_hand.compare_to(h) == 0]